File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def find_good(node, curr_max) -> int:
            if not node:
                return 0
            happy = 0
            if node.val >= curr_max:
                happy = 1
            curr_max = max(curr_max, node.val)
            return happy + find_good(node.left, curr_max) + find_good(node.right, curr_max)
        return find_good(root, root.val)
            

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::vector<int> result{};
        if (!root) return result;
        std::deque<TreeNode*> q{root};
        while (q.size() > 0) {
            int last{};
            int level = q.size();
            for (size_t i{}; i < level; ++i) {
                TreeNode* node = q.front();
                last = node->val;
                if (node->left) q.push_back(node->left);
                if (node->right) q.push_back(node->right);
                q.pop_front();
            }
            result.push_back(last);
        }
        return result;
    }
};
